core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -163,45 +210,88 @@ String getPreferredIp(List<String> preferredTypes, long timeoutMs) {
+   * future that resolves once a valid InstanceData has been acquired. This sets up a chain of
+   * futures that will 1. Acquire a rate limiter. 2. Attempt to fetch instance data. 3. Schedule the
+   * next attempt to get instance data based on the success/failure of this attempt.
+   *
+   * <p>This behavior changed. In version 1.14.1 and earlier, this method was called
+   * performRefresh(). It used to block until either (1) an InstanceData was retrieved or (2) an
+   * attempt failed, and an exception was thrown.
+   *
+   * @see com.google.cloud.sql.core.CloudSqlInstance#handleRefreshResult(
+   *     com.google.common.util.concurrent.ListenableFuture)
    */
-  private InstanceData performRefresh() throws InterruptedException, ExecutionException {
-    logger.fine(
-        String.format("[%s] Refresh Operation: Acquiring rate limiter permit.", instanceName));
+  private ListenableFuture<InstanceData> startRefreshAttempt() {
+    // As soon as we begin submitting refresh attempts to the executor, mark a refresh
+    // as "in-progress" so that subsequent forceRefresh() calls balk until this one completes.
+    synchronized (instanceDataGuard) {
+      refreshRunning = true;
+    }
+
     // To avoid unreasonable SQL Admin API usage, use a rate limit to throttle our usage.
-    //noinspection UnstableApiUsage
-    forcedRenewRateLimiter.acquire();
-    logger.fine(
-        String.format(
-            "[%s] Refresh Operation: Acquired rate limiter permit. Starting refresh...",
-            instanceName));
+    ListenableFuture rateLimit =
+        executor.submit(
+            () -> {
+              logger.fine(
+                  String.format(
+                      "[%s] Refresh Operation: Acquiring rate limiter permit.", instanceName));
+              //noinspection UnstableApiUsage
+              forcedRenewRateLimiter.acquire();
+              logger.fine(
+                  String.format(
+                      "[%s] Refresh Operation: Acquired rate limiter permit. Starting refresh...",
+                      instanceName));
+            },
+            executor);
+
+    // Once rate limiter is done, attempt to getInstanceData.
+    ListenableFuture<InstanceData> dataFuture =
+        Futures.whenAllComplete(rateLimit)
+            .callAsync(
+                () -> {
+                  return instanceDataSupplier.getInstanceData(
+                      this.instanceName,
+                      this.accessTokenSupplier,
+                      this.authType,
+                      executor,
+                      keyPair);
+                },
+                executor);
+
+    // Finally, reschedule refresh after getInstanceData is complete.
+    ListenableFuture<InstanceData> rescheduleFuture =
+        Futures.whenAllComplete(dataFuture)
+            .callAsync(() -> handleRefreshResult(dataFuture), executor);
+
+    return rescheduleFuture;
+  }
 
+  private ListenableFuture<InstanceData> handleRefreshResult(
```

Comment:
Why does this return a value?

Reply:
Because we need to create a chain of futures to retry the refresh when an attempt fails. Let's discuss.

Reply:
Got it -- I didn't understand that callAsync takes a Future.